### DISCUSSION TITLE
Garbage Collector and objects allocation

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#11029] [General]
Hi,
I am trying to decrease the gc duration time (20ms) of the gc runs for an automat looping at 50-100hz. I am missing both understanding of the allocation-deallocation process and analysis tools to investigate. 

_Allocation & de-allocation:_
1. Where are local function objects allocated (for instance float, tuple and list) ? Heap or stack ? 
2. Where are instantiated classes allocated ? Heap ?
3. If so are they part of the sweep process ? 
4. What if an instantiated class persists all along the program life : is this instantiated class walked through each time the gc runs ?
5. In case many classes are instantiated, would it be beneficial to append these classes in a list (kind of “umbrella”), should the gc only sweeps the main (root-umbrella) object, in this case the list ? 
6. Where are objects belonging to an instantiated class allocated (self.x, x being the object, for instance a float) ?
- Are these variables within an instantiated class swept individually, independently of the instantiated class ?
- Or is the instantiated class the “umbrella” and these variables are not swept individually ?

_Tools:_ 
1. Has MicroPython Garbage Collector three generations depending on collection sweeps, as CP has ? 
2. If so, are there any tools to quantify objects broken down per generations ? 
3. Are there any tools to quantify allocated objects broken down per type in the heap ? 

_Hints:_
1. Which process takes significant time within gc: sweeping or de-allocating ?
2. Can these be run separately ?
3. Or can gc be run in chunks ?

Thanks for help.

**Comment by mattytrentini** [comment]
Probably best to start by watching @jimmo's [The MicroPython Garbage Collector](https://www.youtube.com/watch?v=H_xq8IYjh2w) deep dive which will help answer some of those questions.
